docs/zenith/developer/go/457202-test-build-publish.md

```diff
@@ -0,0 +1,303 @@
+
+- Initialize a new Dagger module and add dependencies to it
+- Write custom functions using the Go SDK to test, build and package an application
+- Use containers as function input arguments and return values
+- Debug module code using an interactive shell
+- Test module code by exposing services or containers returned by a function on the host
+
+## Requirements
+
+This guide assumes that:
+
+- You have a good understanding of the JavaScript programming language.
+- You know the basics of programming Dagger modules. If not, refer to the [quickstart](./525021-quickstart.md).
+- You have a Go development environment. If not, install [Go](https://go.dev/doc/install).
+- You have the Dagger CLI installed in your development environment. If not, [install the Dagger CLI](../../../current/cli/979595-reference.md).
+- You have Docker installed and running on the host system. If not, [install Docker](https://docs.docker.com/engine/install/).
+- You have a Node.js Web application. If not, follow the steps in Appendix A to [create an example Node.js application](#appendix-a-create-an-example-application).
+
+## Step 1: Initialize a new module
+
+The example module used in this guide builds, tests and publishes a Node.js application.
+
+Begin by running `dagger mod init` in your application directory to bootstrap a new module:
+
+```sh
+dagger mod init --name=mymod --sdk=go
+```
+
+This will generate a `dagger.json` module file, an initial `main.go` source file, as well as a generated `dagger.gen.go` and `internal` folder for the generated module code.
+
+## Step 2: Add a function to build the application base image
+
+The first step is to add a function to build a base image containing the application source code and runtime. This base image will serve as an input to other functions.
+
+Since the application is a Node.js application, it's convenient to use the `node` module from the [Daggerverse](https://daggerverse.dev), which provides a set of ready-made functions to manage a Node.js project.
+
+1. Add the `node` module as a dependency:
+
+  ```shell
+  dagger mod install github.com/quartz-technology/daggerverse/node@9ce087b83aa8b85f828d7d92ce39bd7c055cfc0f
+  ```
+
+1. Update the generated `main.go` file with the following code:
+
+  ```go
+  package main
+
+  import (
+    "context"
+  )
+
+  type Mymod struct {}
+
+  const defaultNodeVersion = "16"
+
+  func (m *Mymod) buildBase(nodeVersion Optional[string]) *Node {
+    return dag.Node().
+      WithVersion(nodeVersion.GetOr(defaultNodeVersion)).
+      WithNpm().
+      WithSource(dag.Host().Directory(".", HostDirectoryOpts{
+        Exclude: []string{".git", "**/node_modules"},
+      })).
+      Install(nil)
+  }
+  ```
+
+  This function does the following:
+    - It calls the `node` module's `WithVersion()` method via the `dag` client. This method returns a `node` container image with the given Node.js version. This container image is represented as a `Container` object.
```

Comment:
Here and below, the object isn't `Container`, it's actually just called `Node`. This is important to correct since we don't want to imply that users can add functions to core types like `Container`.

Reply:
Modified